Adding reboot action for os_server_action module

gabicrisan opened the pull request:
##### SUMMARY

This enables the reboot action within the os_server_action module. Since there are two types of reboot actions currently supported by Openstack API, a new argument `reboot_type` has been added in order to be able to select between these two options. The action is defaulted to soft reboot.

Reboot states references: https://docs.openstack.org/api-ref/compute/?expanded=reboot-server-reboot-action-detail

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

reboot action for os_server_action module

##### ADDITIONAL INFORMATION

```
TASK [Reboot VM] *********************************************************************************************************************************************************************************************
task path: /ansible/playbooks/test_reboot.yml:50
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: root
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1569413986.96-138881467066555 `" && echo ansible-tmp-1569413986.96-138881467066555="` echo $HOME/.ansible/tmp/ansible-tmp-1569413986.96-138881467066555 `" ) && sleep 0'
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/basic.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/openstack.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/_text.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/parsing/convert_bool.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/parsing/__init__.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/pycompat24.py
Using module_utils file /usr/lib/python2.7/site-packages/ansible/module_utils/six/__init__.py
Using module file /usr/lib/python2.7/site-packages/ansible/modules/cloud/openstack/os_server_action.py
<127.0.0.1> PUT /root/.ansible/tmp/ansible-local-569NxyXFX/tmpvjy2pC TO /root/.ansible/tmp/ansible-tmp-1569413986.96-138881467066555/os_server_action.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1569413986.96-138881467066555/ /root/.ansible/tmp/ansible-tmp-1569413986.96-138881467066555/os_server_action.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'OS_PROJECT_NAME=Redacted OS_USERNAME=Redacted OS_USER_DOMAIN_NAME=Redacted OS_AUTH_URL=Redacted OS_PASSWORD=Redacted OS_PROJECT_DOMAIN_NAME=Redacted /usr/bin/python2 /root/.ansible/tmp/ansible-tmp-1569413986.96-138881467066555/os_server_action.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1569413986.96-138881467066555/ > /dev/null 2>&1 && sleep 0'
changed: [test-vm.internal -> 127.0.0.1] => {
    "changed": true,
    "invocation": {
        "module_args": {
            "action": "reboot",
            "api_timeout": null,
            "auth": null,
            "auth_type": null,
            "availability_zone": null,
            "cacert": null,
            "cert": null,
            "image": null,
            "interface": "public",
            "key": null,
            "region_name": null,
            "server": "test-vm.internal",
            "timeout": 180,
            "verify": null,
            "wait": true
        }
    }
}
```

gabicrisan committed:
Adding action reboot

gabicrisan committed:
updating documentation for os_server_action module

gabicrisan committed:
fix indentation

gabicrisan committed:
fix styling

gabicrisan committed:
fix styling

gabicrisan committed:
adding reboot to argument_spec list

piroux commented on lib/ansible/modules/cloud/openstack/os_server_action.py:
```diff
@@ -239,6 +247,17 @@ def main():
+
+            cloud.compute.post(
+                _action_url(server.id),
+                json={'reboot': {'type': 'HARD'}})
```
Do we really want a default and only "hard" reboot ?

gabicrisan replied:
That's right. Added soft reboot as default and new argument `reboot_type`.

piroux reviewed (changes requested):
> There's also a soft reboot option but that works only when a VM is in an ACTIVE state and that case can be easily covered by the generic reboot module.

Actually a VM might be purposefully unreachable from the controller network, but still manageable from Openstack, so the use-case for soft rebooting is still valid, according to me.

For instance, a module argument like `reboot_type` could be offered and set to `SOFT` by default. In that  case,  the VM state should be ACTIVE beforehand.

gabicrisan committed:
Adding option for soft reboot 

Modified to default the reboot action to `soft` and added `reboot_type` argument to be able to switch to `hard` reboot.

gabicrisan committed:
adding docs references

gabicrisan committed:
update version_added

gabicrisan committed:
fix version_added

cloudnull reviewed (approved):
shipit

sshnaidm commented:
Thanks for submitting patch for Openstack Ansible modules!
We moved Openstack Ansible modules to Openstack repositories.
Next patches should be submitted not with Ansible Github but with
Openstack Gerrit: https://review.opendev.org/#/q/project:openstack/ansible-collections-openstack
Please submit your code there from now.
Thanks for your contribution and sorry for inconvienience.

sshnaidm reviewed (changes requested):
Because we moved Openstack Ansible modules to Openstack repositories, please submit all changes from openstack modules to Openstack Gerrit: https://review.opendev.org/#/q/project:openstack/ansible-collections-openstack Thanks for your contribution and sorry for inconvienience.

jamescassell commented:
@sshnaidm does that require a CLA?  I know openstacksdk requires a CLA to send fixes, so I'm carrying local patches since I can't get the CLA signed.  To date, I've been able to send fixes to the openstack ansible modules since no CLA is required for ansible contributions.

sshnaidm commented:
@jamescassell yes, you are right, all Openstack contributions require License Agreement to be signed.
Please take a look: https://docs.openstack.org/contributors/common/accounts.html

Akasurde commented:
@gabicrisan Thanks for the contribution. Closing as per above.
